React/React hooks.md

```diff
@@ -0,0 +1,754 @@
+
+```jsx
+// BAD
+export const MyComponent: FC = ({ propA, propB }) => {
+  const shape = {
+    a: propA,
+    b: propB,
+  };
+
+  useEffect(() => {
+    /**
+      On each render, `useEffect` is causing an additional re-render,
+      since it's dependency is an object with a new reference every time.
+    */
+    doSomeSideEffectsWithShape(shape);
+  }, [shape])
+
+  // ...
+}
+```
+
+Assuming that `propA` and `propB` are primitive (or memoized, non-primitive) values, with `useMemo` you are ensuring that non-primitive references are retained throughout re-renders, until `propA` or `propB` have changed.
+
+```jsx
+export const MyComponent: FC = ({ propA, propB }) => {
+  useEffect(() => {
+    // Triggers re-renders only if propA and propB have changed
+    const shape = {
+      a: propA,
+      b: propB,
+    };
+
+    doSomeSideEffectsWithShape(shape);
+  }, [propA, propB])
+ 
+  // ...
+}
+```
+
+#### Wrap non-primitive values with useMemo, if you are sending them as props
+
+You can use `useMemo` if you need to pass an non-primitive value as a prop to a component and you know it will cause a large subtree to re-render each time it changes:
+
+```jsx
+export const MyDataChartWithIntersection: FC = ({ xAxisData, yAxisData }) => {
+  // getChartData is expensive calculation
+  const data = useMemo(() => getChartData(xAxisData, yAxisData), [xAxisData, yAxisData]);
+
+  // getIntersections is expensive calculation
+  const intersection = useMemo(() => getIntersections(xAxisData, yAxisData), [xAxisData, yAxisData])
+
+  return (
+    <Chart data={data} intersection={intersection}/>
+  );
+}
+```
+
+#### Do not use `useMemo` as a semantic guarantee that it will be a constant throughout component re-renders
+
+If you need a value to stay the same throughout re-renders, you might think of `useMemo` as a nice way to "mimic" a constant. While it might seem that way, [it is not guaranteed](https://reactjs.org/docs/hooks-reference.html#usememo).
+
+> **You may rely on useMemo as a performance optimization, not as a semantic guarantee.** In the future, React may choose to “forget” some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. Write your code so that it still works without useMemo — and then add it to optimize performance when it's necessary.
+
+```jsx
+// BAD
+export const MyComponent: FC = () => {
+  const id = useMemo(() => generateId(), []); // Not guaranteed to be a constant
+
+  // ...
+}
+```
+
+The best approach is to use lazy initialization:
+
+```jsx
+// GOOD
+export const MyComponent: FC = () => {
+  const [id, _] = useState(() => generateId()); // Initial value will stay the same throughout re-renders
+
+  // ...
+}
+```
+
+You can also store the value in a ref:
+> Note: updating ref values will not trigger a re-render.
+
+```jsx
+type Result<T> = { v: T };
+
+function useConstant<T>(fn: () => T): T {
```

Comment:
I'm thinking to remove `useConstant` example.
 The explanation is here in the thread https://twitter.com/sstur_/status/1384934568285900806
 Maybe I could leave it and write an explanation of what is wrong...
Consider this:
1. render starts
2. you update the ref
3. react concurrent mode has to throw away the work and doesn't commit
You now have a ref that isn't the latest value, it's a value for a render that was never committed.